File: frontend/src/modules/security/repository/userRepository.ts
import type { UserRepository as UserRepo } from '../domain/contract/userContract'
import type { User, UserForm, UserFilter } from '../domain/entity/user'
import { post, get, patch, DEFAULT_API_PATH } from '@/modules/shared/http_handler'
import { getUserRecords, getUserSingleRecords } from './adapters/userRecords'

const loadUsers = async (filter: UserFilter | null): Promise<GridRecord> => {
    const queryFilter = {
        limit: filter?.limit,
        skip: filter?.skip,
        search: filter?.filters.search,
        status: (filter?.filters?.status === undefined || filter?.filters?.status === null)  
                ? null : (filter?.filters?.status ? 1: 0),
    }

    const response = await get(DEFAULT_API_PATH + '/security/users', queryFilter );
    return getUserRecords(response);
}

const saveUser = async (user: UserForm): Promise<User> => {
    const requestData = await post(DEFAULT_API_PATH + '/security/users', user);
    return getUserSingleRecords(requestData);
}

const getUserById = async (userId: string): Promise<User> => {
    const requestData = await get(DEFAULT_API_PATH + '/security/users/' + userId);
    return requestData.data;
}

const changeUserStatus = async (userId: string, status: boolean): Promise<boolean> => {
    const requestData = await patch(DEFAULT_API_PATH + `/security/users/${userId}/${status ? 1 : 0}`, {} );
    return requestData.status === 200;
}

const userExists = async (userName: string): Promise<boolean> => {
    const requestData = await get(DEFAULT_API_PATH + `/security/users/${userName}/exists`, {} );
    return requestData.data.exists;
}

export default {
    loadUsers,
    saveUser,
    getUserById,
    changeUserStatus,
    userExists
} as UserRepo

File: backend/modules/security/repository/user_repository.go
package repository

import (
	securityService "stm/modules/security/domain/contract"
	securityEntity "stm/modules/security/domain/entity"
	sharedModel "stm/shared/model"
	encoding "stm/shared/utility/encoding"
	gormUtil "stm/shared/utility/gorm"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type userRepo struct {
	db *gorm.DB
}

func (l *userRepo) Save(user *securityEntity.User) (securityEntity.User, error) {
	if user.Id == "" {
		user.Id = uuid.New().String()
		user.Password = encoding.GetMD5Hash(user.Password)
	}

	err := l.db.Where(securityEntity.User{Id: user.Id}).
		Assign(user).
		FirstOrCreate(&securityEntity.User{}).Error

	return *user, err
}

func (l *userRepo) GetAllUsers(filter *sharedModel.CriteriaFilter) ([]securityEntity.User, int64, error) {
	users := []securityEntity.User{}

	query := l.db.Model(securityEntity.User{})

	if search, ok := (filter.Filters)["search"]; ok && search != "" {
		query = query.Where("concat(name, ' ', user_name, ' ', last_name) LIKE ?", "%"+search.(string)+"%")
	}

	if isActive, ok := (filter.Filters)["isActive"]; ok && isActive != "" {
		query = query.Where("is_active = ?", isActive)
	}

	// Getting records count
	total := gormUtil.GetCount(l.db, query)

	query = query.Limit(int(filter.Limit))
	query = query.Offset(int(filter.Skip))

	query = query.Order("name desc")
	l.db.Raw("?", query).Scan(&users)

	return users, total, nil
}

func (l *userRepo) GetUserById(userId string) (securityEntity.User, error) {
	user := securityEntity.User{}

	err := l.db.Where(&securityEntity.User{
		Id: userId,
	}).Find(&user).Error

	user.Password = ""

	return user, err
}

func (l *userRepo) ToggleUserStatus(userId string, isActive int8) error {
	user := securityEntity.User{}

	err := l.db.Model(&user).
		Where("id = ?", userId).
		Update("is_active", isActive).Error

	return err
}

func (l *userRepo) UserExists(userName string) (bool, error) {
	var exists bool
	err := l.db.Model(&securityEntity.User{}).
		Select("count(*) > 0").
		Where("user_name = ?", userName).
		Find(&exists).
		Error

	return exists, err
}

func (l *userRepo) ListUsers() ([]securityEntity.User, error) {
	var users = []securityEntity.User{}
	err := l.db.Model(&securityEntity.User{}).
		Find(&users).
		Error

	return users, err
}

func NewUserRepo(db *gorm.DB) securityService.UserRepository {
	return &userRepo{
		db: db,
	}
}
